compiler/frontend/src/mir_optimize/tail_calls.rs

```diff
@@ -0,0 +1,75 @@
+use crate::mir::{Body, Expression, Mir};
+use std::mem;
+
+pub fn simplify_tail_call_tracing(mir: &mut Mir) {
+    // Since this runs as a final pass over the MIR after all other
+    // optimizations, we don't have to update any [`PurenessInsights`].
+    visit_body(&mut mir.body);
+}
+
+fn visit_body(body: &mut Body) {
+    for (_, expression) in &mut body.expressions {
+        if let Expression::Function { body, .. } = expression {
+            visit_body(body);
+            simplify_body(body);
+        }
+    }
+}
+fn simplify_body(body: &mut Body) {
+    let [.., (call_id, Expression::Call { .. }), (
+        _,
+        Expression::TraceCallEnds {
+            return_value: end_return_value,
+        },
+    ), (_, Expression::Reference(reference_target))] = &mut body.expressions[..]
+    else {
+        return;
+    };
```

Comment:
Does this work? Then we wouldn't need the if check below.
```suggestion
    let [.., (end_return_value, Expression::Call { .. }), (
        _,
        Expression::TraceCallEnds {
            return_value: end_return_value,
        },
    ), (_, Expression::Reference(end_return_value))] = &mut body.expressions[..]
    else {
        return;
    };
```